pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/FlattenStep.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.pulsar.functions.transforms;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.common.schema.SchemaType;
+
+@RequiredArgsConstructor
+public class FlattenStep implements TransformStep {
+  // TODO: Cache schema to flattened schema
+  // TODO: Microbenchmark the flatten algorithm for performance optimization
+  // TODO: Validate flatten delimiter
+  // TODO: Add integration test
+
+  private static final String DEFAULT_DELIMITER = "_"; // '.' in not valid in AVRO field names
+
+  private final Optional<String> delimiter;
+  private final Optional<String> part;
+
+  @Override
+  public void process(TransformContext transformContext) throws Exception {
+    if (!part.isPresent()) {
+      validateAvro(transformContext.getKeySchema());
+      validateAvro(transformContext.getValueSchema());
+      GenericRecord avroKeyRecord = (GenericRecord) transformContext.getKeyObject();
+      GenericRecord avroValueRecord = (GenericRecord) transformContext.getValueObject();
+      transformContext.setKeyObject(flattenGenericRecord(avroKeyRecord));
+      transformContext.setValueObject(flattenGenericRecord(avroValueRecord));
+      transformContext.setKeyModified(true);
+      transformContext.setValueModified(true);
+    } else if (part.isPresent() && "key".equals(part.get())) {
```

Comment:
`part.isPresent()` is probably always true here ?

Reply:
you are right. let me change this in-light of the other comment about using optional as class member.